libgit/autogit_node_wrappers.go

```diff
@@ -0,0 +1,174 @@
+// Copyright 2018 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libgit
+
+import (
+	"context"
+
+	"github.com/keybase/kbfs/libkbfs"
+	"github.com/keybase/kbfs/tlf"
+	"github.com/pkg/errors"
+)
+
+// This file contains libkbfs.Node wrappers for implementing the
+// .kbfs_autogit directory structure. It breaks down like this:
+//
+// * `rootWrapper.wrap()` is installed as a root node wrapper, and wraps
+//   the root node for each TLF in a `rootNode` instance.
+// * `rootNode` allows .kbfs_autogit to be auto-created when it is
+//   looked up, and wraps it two ways, as both a `readonlyNode`, and
+//   an `autogitRootNode`.
+// * `readonlyNode` is always marked as read-only, unless
+//   `ctxReadWriteKey` has a non-nil value in the context.
+// * `autogitRootNode` allows the auto-creation of subdirectories
+//   representing TLF types, e.g.. .kbfs_autogit/private or
+//   .kbfs_autogit/public.  It wraps child nodes two ways, as both a
+//   `readonlyNode`, and an `tlfTypeNode`.
+// * `tlfTypeNode` allows the auto-creation of subdirectories
+//   representing TLFs, e.g. .kbfs_autogit/private/max or
+//   .kbfs_autogit/team/keybase.  It wraps child nodes as a
+//   `readOnlyNode`.  TODO(KBFS-2678): allow repo autocreation under
+//   a `tlfTypeNode`.
+
+type ctxReadWriteKeyType int
+
+const (
+	autogitRoot                         = ".kbfs_autogit"
+	ctxReadWriteKey ctxReadWriteKeyType = 1
+
+	public  = "public"
+	private = "private"
+	team    = "team"
+)
+
+// tlfTypeNode represents an autogit subdirectory corresponding to a
+// specific TLF type.  It can only contain subdirectories that
+// correspond to valid TLF name for the TLF type.
+type tlfTypeNode struct {
+	libkbfs.Node
+	am      *AutogitManager
+	tlfType tlf.Type
+}
+
+var _ libkbfs.Node = (*tlfTypeNode)(nil)
+
+// ShouldCreateMissedLookup implements the Node interface for
+// tlfTypeNode.
+func (ttn tlfTypeNode) ShouldCreateMissedLookup(
+	ctx context.Context, name string) (
+	bool, context.Context, libkbfs.EntryType, string) {
+	_, err := libkbfs.ParseTlfHandle(
+		ctx, ttn.am.config.KBPKI(), ttn.am.config.MDOps(), name, ttn.tlfType)
+
+	ctx = context.WithValue(ctx, ctxReadWriteKey, 1)
+	switch e := errors.Cause(err).(type) {
+	case nil:
+		return true, ctx, libkbfs.Dir, ""
+	case libkbfs.TlfNameNotCanonical:
+		return true, ctx, libkbfs.Sym, e.NameToTry
```

Comment:
This would cause the creation of a symlink for the TLF if user-provided TLF name is not canonical right? Seems like it would be very common. Have you considered just disallowing accessing autogit from non-canonical paths?

Reply:
Yes that's right.  I thought this would be way more user-friendly, since it's just like the way /keybase works (except the symlink is permanently to a directory rather than being ephemeral).  Seems pretty useful to me.  What do you think?

Reply:
I was a bit worried if this would end up with a lot of symlinks in `.kbfs_autogit`. But on the other hand now thinking about it more, I suppose user would barely do a `ls` on `.kbfs_autogit` dir, and in GUI we can just filter out the symlinks. So I'm OK with this.